openshift/hac-dev
# HACBS - Application workflow visualization

## Fixes 

https://issues.redhat.com/browse/HACBS-794

## Description


This PR contains the Application Workflow visualization.

The node data model is mapped with the following CRs

- Source node  - > **Component**
- Build  -> Application **PipelineRuns** of type builds
- Component Integration -> **IntegrationTestScenarios** with component `context`
- Application Integration -> **IntegrationTestScenarios** with application `context`
- Static Environments -> **Environments**
- Managed Environments -> **ReleaseLinks**

## Type of change




## Screen shots / Gifs for design review 

Disabled nodes:


![image](https://user-images.githubusercontent.com/9964343/182138348-9bdef191-be42-47bf-85c8-875e8a087b47.png)

Variable width nodes:


![image](https://user-images.githubusercontent.com/9964343/182140975-46ee392f-8af4-4405-88ef-9487bacaefc6.png)


## How to test or reproduce?





1. Create Application, componetns, IntegrationTestScenorios, Envirnoments and releaseLinks CR.

use this [link](https://github.com/karthikjeeyar/hacbs-sample-resources/tree/main/examples) to refer the samples

##Unit Tests:


```
   visualization-utils.spec.ts
  getWorkflowNodes
    ✓ expect to return empty array (2 ms)
    ✓ expect to return same number of nodes for linear workflows
    ✓ expect to contain a spacer node for parallel to parallel workflows (13 ms)
  getTopologyNodesEdges
    ✓ expect to return nodes and edges empty array (1 ms)
    ✓ expect to return nodes and edges for a workflow (2 ms)
  dagtoNodes
    ✓ should return empty for invalid values (1 ms)
    ✓ should return empty array if dag instance does not have vertices (1 ms)
    ✓ should return nodes if dag instance with vertices
  workflowToNodes
    ✓ should return empty array for workflow with no steps
    ✓ should return empty for invalid values (1 ms)
  sortEnvironments
    ✓ should return empty array for invalid values (1 ms)
    ✓ should return sorted array of environments (1 ms)
  getLastEnvironment
    ✓ should return default environment for invalid values (1 ms)
    ✓ should return the last environment in the list```
```


```
 useAppWorkflowData.spec.ts
  ✓ should early return empty values if any one of the resources is not loaded (12 ms)
  ✓ should return the basic workflow nodes (4 ms)
  ✓ should return disabled nodes if the resources are not created (2 ms)
  ✓ Abstract nodes should contain staic name (2 ms)
  ✓ Non Abstract nodes should contain dynamic name (4 ms)
  ✓ nodes should be disabled / enabled based on the availability of resources (2 ms)
  ✓ label should change based on the availability of resources (2 ms)
```

## Browser conformance:
